tests/v3/test_store/test_stateful_store.py

```diff
@@ -0,0 +1,223 @@
+        raise NotImplementedError
+
+    def supports_writes(self):
+        raise NotImplementedError
+
+
+class ZarrStoreStateMachine(RuleBasedStateMachine):
+    keys_bundle = Bundle("keys_bundle")
+    key_ranges_bundle = Bundle("key_ranges")
+
+    def __init__(self):
+        super().__init__()
+        self.model: dict[str, bytes] = {}
+        self.store = SyncStoreWrapper(MemoryStore(mode="w"))
+
+    # NOTE: set this and the next as initialize instead of rule to make sure they run
+    #      and create key, key_ranges before any rules that use key, key_ranges
+    #      not totally sure this is necessary
+    @initialize(key=paths, data=st.binary(min_size=0, max_size=100), target=keys_bundle)
+    def set(self, key: str, data: bytes) -> None:
+        note(f"(set) Setting {key!r} with {data}")
+        assert not self.store.mode.readonly
+        data_buf = Buffer.from_bytes(data)
+        self.store.set(key, data_buf)
+        self.model[key] = data_buf
+        note("init key")
+        return key
+
+    @initialize(key=keys_bundle, data=st.data(), target=key_ranges_bundle)
+    def init_key_ranges(self, data, key):
+        note("init key ranges")
+        key_st = st.just(key)
+        key_ranges_st = key_ranges(keys=key_st)
+        return data.draw(key_ranges_st)
+
+    @precondition(lambda self: len(self.model.keys()) > 0)
+    @rule(key=keys_bundle)
+    def get(self, key) -> None:
+        note(f"(get) {key=}")
+        store_value = self.store.get(key)
+        # to bytes here necessary (on model and store) because data_buf set to model in set()
+        assert self.model[key].to_bytes() == (store_value.to_bytes())
+
+    @precondition(lambda self: len(self.model.keys()) > 0)
+    @rule(key=keys_bundle, key_range=key_ranges_bundle)
+    def get_partial_values(self, key, key_range) -> None:
+        note(f"(get partial) {key_range=}")
+        obs_maybe = self.store.get_partial_values(key_range)
+        observed = []
+
+        for obs in obs_maybe:
+            assert obs is not None
+            observed.append(obs.to_bytes())
+
+        model_vals_ls = []
+
+        for key, byte_range in key_range:
+            model_vals = self.model[key]
+            start = byte_range[0] or 0
+            step = byte_range[1]
+            stop = start + step if step is not None else None
+            assert model_vals[start:stop] is not None
+            model_vals_ls.append(model_vals[start:stop])
+
+        assert all(
+            obs == exp.to_bytes() for obs, exp in zip(observed, model_vals_ls, strict=True)
+        ), (
+            observed,
+            model_vals_ls,
+        )
+
+    @precondition(lambda self: len(self.model.keys()) > 0)
+    @rule(data=st.data())
+    def delete(self, data) -> None:
+        path_st = data.draw(st.sampled_from(sorted(self.model.keys())))
+        note(f"(delete) Deleting {path_st=}")
+
+        self.store.delete(path_st)
+        del self.model[path_st]
```

Comment:
```suggestion
    @rule(key=consumes(keys_bundle))
    def delete(self, data) -> None:
        note(f"(delete) Deleting {key=}")
        self.store.delete(key)
        del self.model[key]
```

https://hypothesis.readthedocs.io/en/latest/stateful.html#hypothesis.stateful.consumes